### PR TITLE
[BugFix] Fix CompoundPredicateOperator not equals when same children with different order

### LINKED ISSUE #8675: query error
### Steps to reproduce the behavior (Required)
```
sql:
select  
  subq_0.c3 as c0, 
  case when cast(nullif(subq_0.c5,
        subq_0.c12) as VARCHAR) = ref_4.t_shift then ref_4.t_minute else ref_4.t_minute end
     as c1, 
  ref_4.t_sub_shift as c2, 
  date_sub(
    cast(subq_0.c9 as DATETIME),
    cast(subq_0.c8 as INT)) as c3, 
  subq_0.c7 as c4
from 
  (select  
          ref_2.d_date as c0, 
          ref_3.w_county as c1, 
          ref_3.w_warehouse_sk as c2, 
          ref_0.p_brand as c3, 
          ref_0.p_color as c4, 
          ref_0.p_name as c5, 
          seconds_sub(
            cast(ref_2.d_date as DATETIME),
            cast(ref_3.w_warehouse_sq_ft as INT)) as c6, 
          ref_0.p_partkey as c7, 
          ref_1.id_int as c8, 
          ref_2.d_date as c9, 
          case when ref_2.d_moy >= ref_2.d_quarter_seq then ref_1.id_datetime else ref_1.id_datetime end
             as c10, 
          ref_1.id_double as c11, 
          ref_0.p_container as c12, 
          ref_3.w_street_name as c13, 
          ref_1.id_datetime as c14, 
          ref_0.p_container as c15, 
          ref_1.id_int as c16, 
          ref_2.d_same_day_ly as c17, 
          ref_0.p_type as c18, 
          98 as c19, 
          ref_1.id_double as c20, 
          ref_2.d_date_id as c21
        from 
          sqlsmith.hive_ssb_1g_csv_part as ref_0
                inner join sqlsmith.test_cross_join_20 as ref_1
                on (ref_0.p_type = ref_1.id_varchar )
              inner join sqlsmith.tpcds_100g_date_dim as ref_2
              on (ref_0.p_color = ref_2.d_date_id )
            left join sqlsmith.tpcds_100g_warehouse as ref_3
            on (ref_2.d_fy_week_seq = ref_3.w_warehouse_sk )
        where ref_1.id_decimal <= ref_3.w_gmt_offset) as subq_0
    left join sqlsmith.tpcds_100g_time_dim as ref_4
    on (subq_0.c3 = ref_4.t_time_id )
where subq_0.c2 > subq_0.c7;
```

```
fe.log:
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:657) ~[?:1.8.0_252]
	at java.util.ArrayList.get(ArrayList.java:433) ~[?:1.8.0_252]
	at com.starrocks.sql.optimizer.Group.getFirstLogicalExpression(Group.java:91) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.rule.Binder.extractGroupExpression(Binder.java:131) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.rule.Binder.match(Binder.java:93) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.rule.Binder.match(Binder.java:93) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.rule.Binder.next(Binder.java:69) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.task.ApplyRuleTask.execute(ApplyRuleTask.java:65) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.task.SeriallyTaskScheduler.executeTasks(SeriallyTaskScheduler.java:42) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.memoOptimize(Optimizer.java:162) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.optimize(Optimizer.java:103) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.StatementPlanner.createQueryPlan(StatementPlanner.java:124) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.StatementPlanner.plan(StatementPlanner.java:89) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:338) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:283) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:401) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:637) ~[starrocks-fe.jar:?]
	at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]

```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - mysql> select current_version();
+-------------------------+
| current_version()       |
+-------------------------+
| BRANCH-2.3-ASAN 5ae3aa9 |
+-------------------------+

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #8675 

## Problem Summary(Required) ：

CompoundPredicateOperator like a=b and c=d not equals c=d and a=b, this will result in operator not equals which use CompoundPredicateOperator as predicate or on predicate

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 9 / 9 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java|9|9|100.00%|[]|

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java [hunk 1]
```diff
@@ -71,16 +78,22 @@ public boolean equals(Object o) {
...
             return false;
         }
-        if (!super.equals(o)) {
+        CompoundPredicateOperator that = (CompoundPredicateOperator) o;
+        if (type != that.type) {
             return false;
         }
-        CompoundPredicateOperator that = (CompoundPredicateOperator) o;
-        return type == that.type;
+        if (type.equals(CompoundType.OR) || type.equals(CompoundType.AND)) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java:85]: not only one child, so needn't check OR and AND.

**Comment by Youngwb** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java|7|7|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Youngwb** [comment]
@Mergifyio backport branch-2.3

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java|7|7|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java|7|7|100.00%|[]|

**Pull request merged**